tools/codegen/api/types.py

```diff
@@ -26,13 +26,63 @@
 # you need trnsnlations that know about these types, beef up this data
 # structure.
 
-@dataclass(frozen=True)
+
+# Prints an enum without the enum class name
+class PrintEnum(Enum):
+    def __str__(self):
+        return self.name
+
+# The set of all non-templated, valid, fully-qualified names of C++ types that are used in the codegen.
+# Templated types get their own dataclass, mainly to make namespace parsing easier.
+BaseCppType = PrintEnum('BaseCppType', (
+    'int64_t',
+    'double',
+    'bool',
+    'void',
+    'std::string',
+    'at::Generator',
+    'at::ScalarType',
+    'at::Tensor',
+    'at::TensorList',
+    'at::Dimname',
+    'at::DimnameList',
+    'at::Layout',
+    'at::Device',
+    'at::Scalar',
+    'at::MemoryFormat',
+    'at::QScheme',
+    'at::Storage',
+    'at::Stream',
+    'at::IntArrayRef',
+    'at::TensorOptions',
+))
+
+@dataclass(unsafe_hash=True)
```

Comment:
I couldn't use `frozen=True` here, since I have a custom `__init__` function :/ added the `unsafe_hash` property instead so we can still use BaseCType as dictionary keys.

Reply:
You can use `frozen=True`, you just have to be a little clever about how you do attribute setting. https://stackoverflow.com/questions/57893902/how-can-i-set-an-attribute-in-a-frozen-dataclass-custom-init-method

But I would examine more closely why you want this to happen in the constructor, since you can always define a helper class method for constructing frozen dataclasses. Maybe that is warranted here?